---
output:
  github_document
---

```{r, echo = FALSE, include = FALSE}
pkg <- read.dcf("DESCRIPTION", fields = "Package")[1]
title <- gsub("\n"," ",read.dcf("DESCRIPTION", fields = "Title")[1])
description <- gsub("\n"," ",read.dcf("DESCRIPTION", fields = "Description")[1])
URL <- read.dcf('DESCRIPTION', fields = 'URL')[1]
owner <- strsplit(URL,"/")[[1]][4]
repo <- strsplit(URL,"/")[[1]][5]
```

# <code>Power Analysis</code><br>Power Analysis for Differential Expression in scRNA-seq data  

`r rworkflows::use_badges(add_hex = FALSE, add_authors = FALSE, add_codecov = FALSE, add_codecov_graphs = FALSE, add_license = FALSE)`
`r badger::badge_license(license = rworkflows::get_description(refs = NULL)[[1]]$get_field("License"))`

**Authors:** ***Salman Fawad, Alan Murphy, Yunning Yuan, Hiranyamaya (Hiru) Dash, Nathan Skene***  
**Updated:** ***`r format(Sys.Date(), '%b-%d-%Y')`***  


## Introduction

The poweranalysis R package is designed to run robust power analysis for differential gene expression in scRNA-seq studies and provides tools to estimate the optimal number of samples and cells needed to achieve reliable power levels.

Wraps work from [this](https://github.com/neurogenomics/scRNA-seq_Power_Analysis) repository into an R package.

To install:  
```R
devtools::install_github("neurogenomics/Power_Analysis")
```

Load:  
```R
library(poweranalysis)
```

## Contact

### [Neurogenomics Lab](https://www.neurogenomics.co.uk)  
UK Dementia Research Institute  
Department of Brain Sciences  
Faculty of Medicine  
Imperial College London  
[GitHub](https://github.com/neurogenomics)  

<hr>

### Session Info

<details>
```{r}
utils::sessionInfo()
```
</details>
