---
title: "Bulk Analysis" 
author: "<h4>Authors: <i>Yunning Yuan, Salman Fawad, Nathan Skene</i></h4>"
date: "<h4>Vignette updated: <i>`r format( Sys.Date(), '%b-%d-%Y')`</i></h4>"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{bulk_anaysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(poweranalysis)
```

## Validating Power Analysis Using Bulk Data
This section introduces a validation approach for scRNA-seq power analysis using high-confidence, sex-biased DEGs from bulk GTEx data. Specifically, it quantifies how many “PTP DEGs” (genes sex-biased in ≥90% of GTEx tissues) are recovered in single-cell DGE results under various down-sampling levels. Since these DEGs are defined across many tissues, they serve as a robust benchmark for sex-bias detection. The `bulk_power_analysis` function estimates power to detect these DEGs in one or more SCE objects, based on metadata specifying how to identify samples and cell types. Below is an example using two scRNA-seq datasets:

```{r bulk-individual, results='hide', message=FALSE, warning=FALSE}
library(poweranalysis)
library(SingleCellExperiment)

# Load SCE objects (replace with actual file paths or data)
tsai_path <- system.file("extdata", "Tsai_Micro.qs", package="poweranalysis")
tsai_micro <- qs::qread(tsai_path)
zhou_path <- system.file("extdata", "Zhou_Astro_subset.qs", package="poweranalysis")
zhou_astro <- qs::qread(zhou_path)

# List of SCE datasets
SCEs <- list(tsai_micro, zhou_astro)

# Dataset names (used in plots and output files)
dataset_names <- c("Tsai_Micro", "Zhou_Astro")

# Cell type mapping
celltype_corr <- list(Micro=c("Micro", NA),
                      Astro=c(NA,"Astro"))

# Metadata column names per SCE
celltypeIDs <- c("cluster_celltype","cluster_celltype")
sampleIDs <- c("sample_id","sample_id")

# Output path (must be a clean directory)
output_indi <- file.path(tempdir(), "bulk_individual_output")
dir.create(output_path, showWarnings = FALSE)

# Load GTEx bulk DEGs
bulk_path <- system.file("extdata", "LFSR.tsv", package="poweranalysis")
bulkDE <- read.table(bulk_path, sep = "\t", header = TRUE)

# Run power analysis for down-sampling individuals
bulk_power_analysis(
  SCEs = SCEs,
  dataset_names = dataset_names,
  celltype_corr = celltype_corr,
  celltypeIDs = celltypeIDs,
  sampleIDs = sampleIDs,
  bulkDE = bulkDE,
  sampled = "individuals",
  output_path = output_indi,
  Nperms = 3 # Nperms=3 for speed in example
)
```

The output plot shows the percentage of PTP DEGs from the bulk data that are recovered in each cell type at various levels of individual down-sampling in the scRNA-seq datasets:

```{r, echo=FALSE, out.width="100%"}
knitr::include_graphics("bulk_individual_output/prop_bulk_DEGs_sc.png")
```

To assess recovery by down-sampling cells instead of individuals, simply set `sampled = "cells"`:

```{r bulk-cell, results='hide', message=FALSE, warning=FALSE}
# Output path (must be a clean directory)
output_cell <- file.path(tempdir(), "bulk_cell_output")
dir.create(output_path, showWarnings = FALSE)

bulk_power_analysis(
  SCEs = SCEs,
  dataset_names = dataset_names,
  celltype_corr = celltype_corr,
  celltypeIDs = celltypeIDs,
  sampleIDs = sampleIDs,
  bulkDE = bulkDE,
  sampled = "cells", 
  output_path = output_cell,
  Nperms = 3
)
```

And here is the corresponding output plot for down-sampling cells:
```{r, echo=FALSE, out.width="100%"}
knitr::include_graphics("bulk_cell_output/prop_bulk_DEGs_sc.png")
```
