---
title: "Bulk Analysis" 
author: "<h4>Authors: <i>Yunning Yuan, Salman Fawad, Nathan Skene</i></h4>"
date: "<h4>Vignette updated: <i>`r format( Sys.Date(), '%b-%d-%Y')`</i></h4>"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Bulk Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE, eval=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(poweranalysis)
```

## Validating Power Analysis using Bulk Data
This section introduces a validation strategy for scRNA-seq power analysis using high-confidence, sex-biased DEGs derived from bulk GTEx data. Specifically, it assesses how many “PTP DEGs” (genes sex-biased in ≥90% of GTEx tissues) are recovered in single-cell differential expression analysis results under various down-sampling levels. Because these DEGs are identified across a broad range of tissues, they provide a robust benchmark for evaluating sex-bias detection. <br>

The `bulk_power_analysis` function estimates the power to detect these DEGs across one or more SCE objects, with user-supplied metadata indicating how to identify samples and cell types. Below is an example that demonstrates this analysis using two scRNA-seq datasets, based on downsampling individuals.

```{r bulk-individual, results='hide', message=FALSE, warning=FALSE, eval=FALSE}
library(poweranalysis)
library(SingleCellExperiment)

# Load SCE objects (replace with actual file paths or data)
tsai_path <- system.file("extdata", "Tsai_Micro.qs", package="poweranalysis")
tsai_micro <- qs::qread(tsai_path)
zhou_path <- system.file("extdata", "Zhou_Astro_subset.qs", package="poweranalysis")
zhou_astro <- qs::qread(zhou_path)

# List of SCE datasets
SCEs <- list(tsai_micro, zhou_astro)

# Dataset names (used in plots and output files)
dataset_names <- c("Tsai_Micro", "Zhou_Astro")

# Cell type mapping
celltype_corr <- list(Micro=c("Micro", NA),
                      Astro=c(NA,"Astro"))

# Metadata column names per SCE
celltypeIDs <- c("cluster_celltype","cluster_celltype")
sampleIDs <- c("sample_id","sample_id")


# Output path (must be a clean directory without any folder)
output_sample <- file.path(getwd(), "bulk_sample")
if (!dir.exists(output_sample)) {
  dir.create(output_sample, recursive = TRUE)
}

# Load GTEx bulk DEGs
bulk_path <- system.file("extdata", "LFSR.tsv", package="poweranalysis")
bulkDE <- read.table(bulk_path, sep = "\t", header = TRUE)

# Run power analysis for down-sampling individuals
bulk_power_analysis(
  SCEs = SCEs,
  dataset_names = dataset_names,
  celltype_corr = celltype_corr,
  celltypeIDs = celltypeIDs,
  sampleIDs = sampleIDs,
  bulkDE = bulkDE,
  sampled = "individuals",
  output_path = output_sample,
  Nperms = 3 # Nperms=3 for speed in example
)
```

To assess recovery by down-sampling cells instead of individuals, simply set `sampled = "cells"`:

```{r bulk-cell, results='hide', message=FALSE, warning=FALSE, eval=FALSE}
# Output path (must be a clean directory)
output_cell <- file.path(getwd(), "bulk_cell")
if (!dir.exists(output_cell)) {
  dir.create(output_cell, recursive = TRUE)
}

bulk_power_analysis(
  SCEs = SCEs,
  dataset_names = dataset_names,
  celltype_corr = celltype_corr,
  celltypeIDs = celltypeIDs,
  sampleIDs = sampleIDs,
  bulkDE = bulkDE,
  sampled = "cells",  # down-sampling cells
  output_path = output_cell,
  Nperms = 3
)
```


The **output** displays the percentage of PTP DEGs from the bulk data that are detected across all cell types at different down-sampling levels of the scRNA-seq datasets. <br>

Down-sampling individuals: <br>
```{r, echo=FALSE, out.width="100%", eval=FALSE}
knitr::include_graphics("bulk_sample/prop_bulk_DEGs_sc.png")
```

Down-sampling cells: <br>
```{r, echo=FALSE, out.width="100%", eval=FALSE}
knitr::include_graphics("bulk_cell/prop_bulk_DEGs_sc.png")
```
