---
title: "Get Started" 
author: "<h4>Authors: <i>Salman Fawad, Yunning Yuan, Alan Murphy, Nathan Skene</i></h4>"
date: "<h4>Vignette updated: <i>`r format( Sys.Date(), '%b-%d-%Y')`</i></h4>"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{getting_started}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r, echo=FALSE, include=FALSE}
pkg <- read.dcf("../DESCRIPTION", fields = "Package")[1]
library(pkg, character.only = TRUE)
```

The *poweranalysis* R package is designed to run robust power analysis for differential gene expression in scRNA-seq studies and provides tools to estimate the optimal number of samples and cells needed to achieve reliable power levels.

## Setup
```{r setup}
library(poweranalysis)
```

## Differential Gene Expression (DGE) Analysis
Import an SCE object and perform differential expression analysis using a pseudobulking approach, enabling the robust identification of differentially expressed genes (DEGs) across conditions or groups from single-cell data. <br>
**Example Usage**: <br>
To run the DGE analysis, first load your own SingleCellExperiment (SCE) object.

```{r, message=FALSE, warning=FALSE}
# Load your SCE object (replace with actual file path)
library(qs)
library(SingleCellExperiment)

SCE <- qs::qread("./data/sce.qs")
```

To use the `DGE_analysis` function, specify the formula for comparison along with the pseudobulk ID and celltype ID. The function requires the following key arguments:

· **`design`**: A formula that defines the variables included in the model for differential expression analysis. This determines how gene expression is compared across different groups.

· **`coef`**: A character string that specifies which group in the `design` formula you want to investigate for differential expression. <br>

For example, to validate the differential expression (DEG) analysis approach, you can run a comparison between sexes using the <i>formula = ~ sex. </i>. This will assess how gene expression differs between male and female groups. 

```{r}
# Run the DGE_analysis function for a sex comparison
DGE_analysis.sex <- DGE_analysis(
  SCE,
  design = ~ sex,
  pseudobulk_ID = "sample_id",
  celltype_ID = "cluster_celltype",
  coef = "M"
)
```

If you want to compare disease and control conditions, specify the disease status in the formula and the disease group of interest in the coef.

```{r}
# Run the DGE_analysis function for a disease vs. control comparison
DGE_analysis.AD_sex <- DGE_analysis(
  SCE,
  design = ~ pathological_diagnosis,
  pseudobulk_ID = "sample_id",
  celltype_ID = "cluster_celltype",
  coef = "AD"
)
```

