---
title: "Getting Started"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{getting_started}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
***
# Introduction

The *poweranalysis* R package is designed to run robust power analysis for differential gene expression in scRNA-seq studies, addressing the challenge of low statistical power due to small sample sizes. It provides tools to estimate the optimal number of samples and cells needed to achieve reliable power levels.
<br>
<br>
Using *poweranalysis* involves four steps: <br>
**1. Differential Expression (DE) Analysis:** Import your SCE object and perform DE analysis using a pseudobulking approach, allowing for the robust identification of differentially expressed genes (DEGs) across conditions or groups. <br>
**2. Correlation Analysis:** Assess the consistency of DEG effect sizes across random permutations, subsets of your data, and different studies by computing correlation matrices. <br>
**3. Downsampling and Power Analysis:** Test the robustness of your findings by downsampling both individuals and cells, then analyze the effect on DEG detection power. <br>
**4. Visualisation:** Generate comprehensive power plots and visual summaries of the power analysis results, including DEG detection rates, false discovery rates, and mean LogFC correlation between down-sampled subsets. 
<br>

# Setup

```{r setup}
library(poweranalysis)
```
