<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Downsample the dataset, based either on the individuals or cells, and run DE analysis on each downsampled output. Save results in a dataframe — downsampling_DEanalysis • poweranalysis</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Downsample the dataset, based either on the individuals or cells, and run DE analysis on each downsampled output. Save results in a dataframe — downsampling_DEanalysis"><meta property="og:description" content="Downsample the dataset, based either on the individuals or cells, and run DE analysis on each downsampled output. Save results in a dataframe"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">poweranalysis</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/getting_started.html">Get Started</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/neurogenomics/Power_Analysis_package/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Downsample the dataset, based either on the individuals or cells, and run DE analysis on each downsampled output. Save results in a dataframe</h1>
    <small class="dont-index">Source: <a href="https://github.com/neurogenomics/Power_Analysis_package/blob/Yunning/R/downsampling_DEanalysis.r" class="external-link"><code>R/downsampling_DEanalysis.r</code></a></small>
    <div class="hidden name"><code>downsampling_DEanalysis.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Downsample the dataset, based either on the individuals or cells, and run DE analysis on each downsampled output. Save results in a dataframe</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">downsampling_DEanalysis</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  range_downsampled <span class="op">=</span> <span class="st">"placeholder"</span>,</span>
<span>  output_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  sampled <span class="op">=</span> <span class="st">"individuals"</span>,</span>
<span>  sampleID <span class="op">=</span> <span class="st">"donor_id"</span>,</span>
<span>  design <span class="op">=</span> <span class="st">"placeholder"</span>,</span>
<span>  sexID <span class="op">=</span> <span class="st">"sex"</span>,</span>
<span>  celltypeID <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>  coeff <span class="op">=</span> <span class="st">"male"</span>,</span>
<span>  fdr <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  nom_pval <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  Nperms <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  y <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  region <span class="op">=</span> <span class="st">"single_region"</span>,</span>
<span>  control <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pval_adjust_method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  rmv_zero_count_genes <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>the input data (should be an SCE object)</p></dd>


<dt id="arg-range-downsampled">range_downsampled<a class="anchor" aria-label="anchor" href="#arg-range-downsampled"></a></dt>
<dd><p>vector or list containing values which the data will be downsampled at, in ascending order</p></dd>


<dt id="arg-output-path">output_path<a class="anchor" aria-label="anchor" href="#arg-output-path"></a></dt>
<dd><p>base path in which outputs will be stored</p></dd>


<dt id="arg-sampled">sampled<a class="anchor" aria-label="anchor" href="#arg-sampled"></a></dt>
<dd><p>downsampling carried out based on what (either "individuals" or "cells")</p></dd>


<dt id="arg-sampleid">sampleID<a class="anchor" aria-label="anchor" href="#arg-sampleid"></a></dt>
<dd><p>sample ID</p></dd>


<dt id="arg-design">design<a class="anchor" aria-label="anchor" href="#arg-design"></a></dt>
<dd><p>the design formula of class type <code>formula</code>. Equation used to fit the model- data for the generalised linear model e.g. expression ~ sex + pmi + disease</p></dd>


<dt id="arg-sexid">sexID<a class="anchor" aria-label="anchor" href="#arg-sexid"></a></dt>
<dd><p>sex ID</p></dd>


<dt id="arg-celltypeid">celltypeID<a class="anchor" aria-label="anchor" href="#arg-celltypeid"></a></dt>
<dd><p>cell type ID</p></dd>


<dt id="arg-coeff">coeff<a class="anchor" aria-label="anchor" href="#arg-coeff"></a></dt>
<dd><p>which coefficient to carry out DE analysis with respect to</p></dd>


<dt id="arg-fdr">fdr<a class="anchor" aria-label="anchor" href="#arg-fdr"></a></dt>
<dd><p>the cut-off False Discovery Rate below which to select DEGs</p></dd>


<dt id="arg-nom-pval">nom_pval<a class="anchor" aria-label="anchor" href="#arg-nom-pval"></a></dt>
<dd><p>the cut-off nominal P-value below which to select DEGs (as an alternative to FDR)</p></dd>


<dt id="arg-nperms">Nperms<a class="anchor" aria-label="anchor" href="#arg-nperms"></a></dt>
<dd><p>number of subsets created when downsampling at each level</p></dd>


<dt id="arg-y">y<a class="anchor" aria-label="anchor" href="#arg-y"></a></dt>
<dd><p>the column name in the SCE object for the return variable e.g. "diagnosis" - Case or disease. Default is the last variable in the design formula. y can be discrete (logistic regression) or continuous (linear regression)</p></dd>


<dt id="arg-region">region<a class="anchor" aria-label="anchor" href="#arg-region"></a></dt>
<dd><p>the column name in the SCE object for the study region. If there are multiple regions in the study (for example two brain regions). Pseudobulk values can be derived separately. Default is "single_region" which will not split by region.</p></dd>


<dt id="arg-control">control<a class="anchor" aria-label="anchor" href="#arg-control"></a></dt>
<dd><p>character specifying which control level for the differential expression analysis e.g. in a case/control/other study use "control" in the y column to compare against. NOTE only need to specify if more than two groups in y, leave as default value for two groups or continuous y. Default is NULL.</p></dd>


<dt id="arg-pval-adjust-method">pval_adjust_method<a class="anchor" aria-label="anchor" href="#arg-pval-adjust-method"></a></dt>
<dd><p>the adjustment method for the p-value in the differential expression analysis. Default is benjamini hochberg "BH". See  stats::p.adjust for available options</p></dd>


<dt id="arg-rmv-zero-count-genes">rmv_zero_count_genes<a class="anchor" aria-label="anchor" href="#arg-rmv-zero-count-genes"></a></dt>
<dd><p>whether genes with no count values in any cell should be removed. Default is TRUE
Saves all DGE analysis outputs for downsampled files as well as a summary table of results showing number of true DEGs detected at each number of samples/cells</p></dd>

</dl></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Salman Fawad, Yunning Yuan.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

