<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Package index â€¢ poweranalysis</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">poweranalysis</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/bulk_analysis.html">Bulk Analysis</a>
    </li>
    <li>
      <a href="../articles/correlation_analysis.html">Correlation Analysis</a>
    </li>
    <li>
      <a href="../articles/power_analysis.html">Get Started</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/neurogenomics/Power_Analysis_package/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="DGE_analysis.html">DGE_analysis()</a></code> </p>
        </td>
        <td><p>Perform differential expression analysis using a pseudobulk approach on single-cell data</p></td>
      </tr><tr><td>
          <p><code><a href="bulk_downsampling_DGEanalysis.html">bulk_downsampling_DGEanalysis()</a></code> </p>
        </td>
        <td><p>Downsample the dataset, based either on the individuals or cells, and run DE analysis on each downsampled output. Save results in a dataframe</p></td>
      </tr><tr><td>
          <p><code><a href="bulk_downsampling_range.html">bulk_downsampling_range()</a></code> </p>
        </td>
        <td><p>Obtain the range of values to downsample at (either for individuals, or mean number of cells per individual), for bulk analysis</p></td>
      </tr><tr><td>
          <p><code><a href="bulk_power_analysis.html">bulk_power_analysis()</a></code> </p>
        </td>
        <td><p>Perform cross-dataset power analysis using scRNA-seq and bulk RNA-seq DEG overlap</p></td>
      </tr><tr><td>
          <p><code><a href="compute_downsampled_corr.html">compute_downsampled_corr()</a></code> </p>
        </td>
        <td><p>For a given down-sampled DGE analysis output, computes the correlation of the log-foldchange of the DEGs (at specified p-value) for a given dataset (celltype)</p></td>
      </tr><tr><td>
          <p><code><a href="correlation_analysis.html">correlation_analysis()</a></code> </p>
        </td>
        <td><p>Perform correlation analysis of DEG effect sizes between single-cell datasets</p></td>
      </tr><tr><td>
          <p><code><a href="correlation_boxplots.html">correlation_boxplots()</a></code> </p>
        </td>
        <td><p>Obtain box plots for the correlations of all celltypes, and the mean correlations at a specified cutoff p-value</p></td>
      </tr><tr><td>
          <p><code><a href="differential_expression.html">differential_expression()</a></code> </p>
        </td>
        <td><p>Differential Expression Analysis using edgeR LRT on pseudobulk data</p></td>
      </tr><tr><td>
          <p><code><a href="downsampling_DEanalysis.html">downsampling_DEanalysis()</a></code> </p>
        </td>
        <td><p>Downsample the dataset, based either on the individuals or cells, and run DE analysis on each downsampled output. Save results in a dataframe</p></td>
      </tr><tr><td>
          <p><code><a href="downsampling_corrplots.html">downsampling_corrplots()</a></code> </p>
        </td>
        <td><p>Create correlation plots of the effect sizes of the top 1000 and top 500 DEGs</p></td>
      </tr><tr><td>
          <p><code><a href="downsampling_range.html">downsampling_range()</a></code> </p>
        </td>
        <td><p>Obtain the range of values to downsample at (either for individuals, or mean number of cells per individual)</p></td>
      </tr><tr><td>
          <p><code><a href="gather_celltype_DEGs.html">gather_celltype_DEGs()</a></code> </p>
        </td>
        <td><p>Collate DEGs detected in DGE analysis outputs, across all celltypes in a dataset (datasets/DGE analysis outputs should have common celltypes as specified below)</p></td>
      </tr><tr><td>
          <p><code><a href="make_pseudobulk.html">make_pseudobulk()</a></code> </p>
        </td>
        <td><p>Calculate the summed pseudobulk values for an SCE object based on one single cell type only. Ensure to filter SCE to pass one cell type's data.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_celltype_correlation.html">plot_celltype_correlation()</a></code> </p>
        </td>
        <td><p>Computes the correlation of the log-foldchange of the DEGs (at specified p-value) for a given celltype, across all datasets</p></td>
      </tr><tr><td>
          <p><code><a href="plot_de_analysis.html">plot_de_analysis()</a></code> </p>
        </td>
        <td><p>Create differential expression analysis plots. Run by DGE_analysis()</p></td>
      </tr><tr><td>
          <p><code><a href="plot_mean_correlation.html">plot_mean_correlation()</a></code> </p>
        </td>
        <td><p>Obtain the average correlation (across celltypes) at a specified cutoff p-value</p></td>
      </tr><tr><td>
          <p><code><a href="power_analysis.html">power_analysis()</a></code> </p>
        </td>
        <td><p>Perform robust power analysis for differential gene expression in scRNA-seq dataset</p></td>
      </tr><tr><td>
          <p><code><a href="power_plots.html">power_plots()</a></code> </p>
        </td>
        <td><p>Create plots for power analysis, with down-sampling based either on the individuals or cells</p></td>
      </tr><tr><td>
          <p><code><a href="preliminary_plots.html">preliminary_plots()</a></code> </p>
        </td>
        <td><p>Create preliminary plots for data exploration</p></td>
      </tr><tr><td>
          <p><code><a href="prop_bulk_DEGs_sc.html">prop_bulk_DEGs_sc()</a></code> </p>
        </td>
        <td><p>Obtain overall percentage overlap between DEGs from bulk data (DE across all tissues) and various scRNA-seq datasets, across all common cell types</p></td>
      </tr><tr><td>
          <p><code><a href="random_permutations.html">random_permutations()</a></code> </p>
        </td>
        <td><p>Obtain highly randomised permutations of a specified dataset (based on sex labels), while maintaining consistency with the sample ID</p></td>
      </tr><tr><td>
          <p><code><a href="sample_cells.html">sample_cells()</a></code> </p>
        </td>
        <td><p>Sample the data for a specified mean number of cells (per sample) from the dataset</p></td>
      </tr><tr><td>
          <p><code><a href="sample_individuals.html">sample_individuals()</a></code> </p>
        </td>
        <td><p>Sample the data for a specified number of individuals from the dataset</p></td>
      </tr><tr><td>
          <p><code><a href="sex_chromosome_DEGs.html">sex_chromosome_DEGs()</a></code> </p>
        </td>
        <td><p>Given a DGE analysis output, this outputs a subset of the analysis with the genes laying only on the sex chromosomes (X/Y)</p></td>
      </tr><tr><td>
          <p><code><a href="subset_pairs.html">subset_pairs()</a></code> </p>
        </td>
        <td><p>Obtain independent pairs of subsets of a specified dataset, based on sample ID</p></td>
      </tr><tr><td>
          <p><code><a href="validate_input_parameters_bulk.html">validate_input_parameters_bulk()</a></code> </p>
        </td>
        <td><p>Tests input parameters for functions</p></td>
      </tr><tr><td>
          <p><code><a href="validate_input_parameters_correlation.html">validate_input_parameters_correlation()</a></code> </p>
        </td>
        <td><p>Tests input parameters for functions</p></td>
      </tr><tr><td>
          <p><code><a href="validate_input_parameters_de.html">validate_input_parameters_de()</a></code> </p>
        </td>
        <td><p>Validate that there are no user input errors for the differential expression analysis - sc.cell.type.de</p></td>
      </tr><tr><td>
          <p><code><a href="validate_input_parameters_power.html">validate_input_parameters_power()</a></code> </p>
        </td>
        <td><p>Tests input parameters for functions</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Salman Fawad, Alan Murphy, Yunning Yuan, Nathan Skene.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

