<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Get Started • poweranalysis</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Get Started">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">poweranalysis</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/power_analysis.html">Get Started</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/neurogenomics/Power_Analysis_package/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Get Started</h1>
                        <h4 data-toc-skip class="author"><h4>
<br>
Authors: <i>Salman Fawad, Yunning Yuan, Alan Murphy, Nathan
Skene</i><br>
</h4></h4>
            
            <h4 data-toc-skip class="date"><h4>
<br>
Vignette updated: <i>Apr-28-2025</i><br>
</h4></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/neurogenomics/Power_Analysis_package/blob/Yunning/vignettes/power_analysis.Rmd" class="external-link"><code>vignettes/power_analysis.Rmd</code></a></small>
      <div class="hidden name"><code>power_analysis.Rmd</code></div>

    </div>

    
    
<p>The <em>poweranalysis</em> R package is designed to run robust power
analysis for differential gene expression in scRNA-seq studies and
provides tools to estimate the optimal number of samples and cells
needed to achieve reliable power levels.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/neurogenomics/Power_Analysis_package" class="external-link">poweranalysis</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="differential-gene-expression-dge-analysis">Differential Gene Expression (DGE) Analysis<a class="anchor" aria-label="anchor" href="#differential-gene-expression-dge-analysis"></a>
</h2>
<p>Import an SCE object and perform differential expression analysis
using a pseudobulking approach, enabling the robust identification of
differentially expressed genes (DEGs) across conditions or groups from
single-cell data. <br> To run the DGE analysis, first load your own
SingleCellExperiment (SCE) object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load your SCE object (replace with actual file path)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/qsbase/qs" class="external-link">qs</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">SCE</span> <span class="op">&lt;-</span> <span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qread.html" class="external-link">qread</a></span><span class="op">(</span><span class="st">"sce.qs"</span><span class="op">)</span></span></code></pre></div>
<p>To use the <code>DGE_analysis</code> function, specify the formula
for comparison along with the pseudobulk ID and celltype ID. The
function requires the following key arguments:</p>
<p>· <strong><code>design</code></strong>: A formula that defines the
variables included in the model for differential expression analysis.
This determines how gene expression is compared across different
groups.</p>
<p>· <strong><code>coef</code></strong>: A character string that
specifies which group in the <code>design</code> formula you want to
investigate for differential expression. <br></p>
<p><strong>Example Usage</strong>: <br> To validate the differential
expression (DEG) analysis approach, you can run a comparison between
sexes using the <i>design = ~ sex. </i> This will assess how gene
expression differs between male and female groups.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run the DGE_analysis function for a sex comparison</span></span>
<span><span class="va">DGE_analysis.sex</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DGE_analysis.html">DGE_analysis</a></span><span class="op">(</span></span>
<span>  <span class="va">SCE</span>,</span>
<span>  design <span class="op">=</span> <span class="op">~</span> <span class="va">sex</span>,</span>
<span>  coef <span class="op">=</span> <span class="st">"M"</span>,</span>
<span>  celltypeID<span class="op">=</span><span class="st">"cluster_celltype"</span>,</span>
<span>  sampleID <span class="op">=</span> <span class="st">"sample_id"</span>,</span>
<span><span class="op">)</span></span></code></pre></div>
<p>If you want to compare disease and control conditions, specify the
disease status in the formula and the disease group of interest in the
coef.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run the DGE_analysis function for a disease vs. control comparison</span></span>
<span><span class="va">DGE_analysis.AD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DGE_analysis.html">DGE_analysis</a></span><span class="op">(</span></span>
<span>  <span class="va">SCE</span>,</span>
<span>  design <span class="op">=</span> <span class="op">~</span> <span class="va">pathological_diagnosis</span>,</span>
<span>  coef <span class="op">=</span> <span class="st">"AD"</span>,</span>
<span>  sampleID <span class="op">=</span> <span class="st">"sample_id"</span>,</span>
<span>  celltypeID <span class="op">=</span> <span class="st">"cluster_celltype"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="power-analysis">Power Analysis<a class="anchor" aria-label="anchor" href="#power-analysis"></a>
</h2>
<p>Perform power analysis to estimate the accuracy and reliability of
DEG detection in your scRNA‑seq dataset under different levels of
sampling. This evaluates how well DEGs can be recovered at varying
numbers of individuals and cells. DEGs identified in each down‑sampled
subset are compared to those from the full dataset to compute the
percentage of true positives recovered, along with the False Discovery
Rate (FDR).</p>
<p>To assess power based on sex‑specific DEGs, use the following
function:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Specify the down-sampling range (or use the default range)</span></span>
<span><span class="va">range_ind</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">20</span>,<span class="fl">30</span>,<span class="fl">40</span><span class="op">)</span></span>
<span><span class="va">range_cell</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">30</span>,<span class="fl">50</span>,<span class="fl">70</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run the power_analysis function for a sex comparison</span></span>
<span><span class="va">power_analysis.sex</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/power_analysis.html">power_analysis</a></span><span class="op">(</span></span>
<span>  <span class="va">SCE</span>,</span>
<span>  range_downsampled_individuals <span class="op">=</span> <span class="va">range_ind</span>,  </span>
<span>  range_downsampled_cells <span class="op">=</span> <span class="va">range_cell</span>,</span>
<span>  design <span class="op">=</span> <span class="op">~</span> <span class="va">sex</span>,</span>
<span>  coef <span class="op">=</span> <span class="st">"M"</span>,</span>
<span>  sampleID <span class="op">=</span> <span class="st">"sample_id"</span>,</span>
<span>  celltypeID <span class="op">=</span> <span class="st">"cluster_celltype"</span>,</span>
<span>  Nperms <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p>The <code>power_analysis</code> function generates several key
outputs:</p>
<ul>
<li><p><strong>QC plots</strong> display distributions of effect sizes
(log2 fold-change) across detected DEGs and the number of cells per
individual in the full dataset.</p></li>
<li><p><strong>DGE analysis results</strong> identify PTP DEGs and
non-DEGs using a 0.05 cut-off for both nominal and adjusted p-values
over the down-sampling range of datasets.</p></li>
<li><p><strong>Power plots</strong> show the mean percentage of PTP DEGs
detected and FDR trends as sample size or number of cells per sample
increases. <br> Down-sampling individuals: <br></p></li>
</ul>
<p>Down-sampling cells: <br></p>
<ul>
<li>
<strong>Effect size-specific detection rates</strong> assess the DEG
recovery across different absolute log2 fold-change bins. <br>
Down-sampling individuals: <br>
</li>
</ul>
<p>Down-sampling cells: <br></p>
<ul>
<li>
<strong>Correlation plots</strong> assess the correlation of effect
sizes for the top 500 and 1000 genes across down-sampling levels. <br>
Down-sampling individuals: <br>
</li>
</ul>
<p>Down-sampling cells: <br></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Salman Fawad, Alan Murphy, Yunning Yuan, Nathan Skene.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
