<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Validation and Evaluation • poweranalysis</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Validation and Evaluation">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">poweranalysis</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/power_analysis.html">Get Started</a>
    </li>
    <li>
      <a href="../articles/validation_analysis.html">Validation and Evaluation</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/neurogenomics/Power_Analysis_package/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Validation and Evaluation</h1>
                        <h4 data-toc-skip class="author"><h4>
<br>
Authors: <i>Yunning Yuan, Salman Fawad, Nathan Skene</i><br>
</h4></h4>
            
            <h4 data-toc-skip class="date"><h4>
<br>
Vignette updated: <i>Jun-02-2025</i><br>
</h4></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/neurogenomics/Power_Analysis_package/blob/master/vignettes/validation_analysis.Rmd" class="external-link"><code>vignettes/validation_analysis.Rmd</code></a></small>
      <div class="hidden name"><code>validation_analysis.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/neurogenomics/Power_Analysis_package" class="external-link">poweranalysis</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="validating-power-analysis-using-bulk-data">Validating Power Analysis using Bulk Data<a class="anchor" aria-label="anchor" href="#validating-power-analysis-using-bulk-data"></a>
</h2>
<p>This section introduces a validation approach for scRNA-seq power
analysis using high-confidence, sex-biased DEGs from bulk GTEx data. The
goal is to quantify what fraction of “PTP DEGs”(genes consistently
sex-biased in ≥90% of GTEx tissues) can be recovered in your single-cell
DGE results across various down-sampling levels of individuals or cells.
Since these DEGs are defined across many tissues, they serve as a robust
benchmark for assessing sex-bias detectability. <br></p>
<p>The <code>bulk_power_analysis</code> function takes as input one or
more <code>SCE</code> objects, along with metadata that defines how cell
types and samples are labeled in each dataset. Using a provided bulk DEG
reference (i.e., GTEx data), the function estimates statistical power to
detect these DEGs under user-defined down-sampling schemes. Below is an
example that demonstrates how to run <code><a href="../reference/bulk_power_analysis.html">bulk_power_analysis()</a></code>
with two scRNA-seq datasets:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/neurogenomics/Power_Analysis_package" class="external-link">poweranalysis</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load SCE objects (replace with actual file paths or data)</span></span>
<span><span class="va">allen_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Allen_Endo_subset.qs"</span>, package<span class="op">=</span><span class="st">"poweranalysis"</span><span class="op">)</span></span>
<span><span class="va">allen_endo</span> <span class="op">&lt;-</span> <span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qread.html" class="external-link">qread</a></span><span class="op">(</span><span class="va">allen_path</span><span class="op">)</span></span>
<span><span class="va">tsai_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Tsai_Micro.qs"</span>, package<span class="op">=</span><span class="st">"poweranalysis"</span><span class="op">)</span></span>
<span><span class="va">tsai_micro</span> <span class="op">&lt;-</span> <span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qread.html" class="external-link">qread</a></span><span class="op">(</span><span class="va">tsai_path</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># List of SCE datasets</span></span>
<span><span class="va">SCEs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">allen_endo</span>, <span class="va">tsai_micro</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Dataset names (used in plots and output files)</span></span>
<span><span class="va">dataset_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Allen_Endo"</span>, <span class="st">"Tsai_Micro"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Cell type mapping</span></span>
<span><span class="va">celltype_corr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Endo<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cerebral cortex endothelial cell"</span>,<span class="cn">NA</span><span class="op">)</span>,</span>
<span>                      Micro<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>,<span class="st">"Micro"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Metadata column names per SCE</span></span>
<span><span class="va">celltypeIDs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell_type"</span>,<span class="st">"cluster_celltype"</span><span class="op">)</span></span>
<span><span class="va">sampleIDs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"donor_id"</span>,<span class="st">"sample_id"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load GTEx bulk DEGs</span></span>
<span><span class="va">bulk_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"LFSR.tsv"</span>, package<span class="op">=</span><span class="st">"poweranalysis"</span><span class="op">)</span></span>
<span><span class="va">bulkDE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">bulk_path</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run power analysis (Nperms=3 for speed in example)</span></span>
<span><span class="fu"><a href="../reference/bulk_power_analysis.html">bulk_power_analysis</a></span><span class="op">(</span></span>
<span>  SCEs <span class="op">=</span> <span class="va">SCEs</span>,</span>
<span>  dataset_names <span class="op">=</span> <span class="va">dataset_names</span>,</span>
<span>  celltype_corr <span class="op">=</span> <span class="va">celltype_corr</span>,</span>
<span>  celltypeIDs <span class="op">=</span> <span class="va">celltypeIDs</span>,</span>
<span>  sampleIDs <span class="op">=</span> <span class="va">sampleIDs</span>,</span>
<span>  bulkDE <span class="op">=</span> <span class="va">bulkDE</span>,</span>
<span>  sampled <span class="op">=</span> <span class="st">"individuals"</span>,</span>
<span>  Nperms <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The output plot displays the percentage of PTP DEGs from the bulk
data that are detected across all cell types at different down-sampling
levels of the scRNA-seq datasets:</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Salman Fawad, Alan Murphy, Yunning Yuan, Nathan Skene.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
